core/arch/arm/sta/interrupt_tests.c

```diff
@@ -0,0 +1,264 @@
+			       TEE_Param params[4] __unused,
+			       void **ppsess __unused)
+{
+	return TEE_SUCCESS;
+}
+
+static void close_session(void *psess __unused)
+{
+}
+
+static size_t test_sgi_value[8];
+static size_t test_spi_value[8];
+static size_t test_ppi_value[8];
+static size_t expect_sgi_value[8];
+static size_t expect_spi_value[8];
+static size_t expect_ppi_value[8];
+
+static enum itr_return __maybe_unused ihandler_ok(struct itr_handler *handler)
+{
+	size_t core_num = get_core_pos();
+
+	assert(core_num < 8);
+
+	if (handler->it < SGI_NUM)
+		test_sgi_value[core_num]++;
+	else if (handler->it < PPI_NUM)
+		test_ppi_value[core_num]++;
+	else
+		test_spi_value[core_num]++;
+
+	DMSG("I'm core [%zu].\n", core_num);
+	DMSG("I have handled the interrupt[%zu].\n", handler->it);
+	return ITRR_HANDLED;
+}
+
+struct itr_handler sgi_handler = {
+	.it = TEST_SGI_ID,
+	.handler = ihandler_ok,
+};
+
+struct itr_handler spi_handler = {
+	.it = TEST_SPI_ID,
+	.handler = ihandler_ok,
+};
+
+struct itr_handler ppi_handler = {
+	.it = TEST_PPI_ID,
+	.handler = ihandler_ok,
+};
+
+static TEE_Result test_sgi(void)
+{
+	itr_add(&sgi_handler);
+	itr_enable(TEST_SGI_ID);
+
+	expect_sgi_value[0]++;
+	expect_sgi_value[1]++;
+	expect_sgi_value[2]++;
+	expect_sgi_value[3]++;
+	itr_raise_sgi(TEST_SGI_ID, 0x1 << 0);
+	itr_raise_sgi(TEST_SGI_ID, 0x1 << 1);
+	itr_raise_sgi(TEST_SGI_ID, 0x1 << 2);
+	itr_raise_sgi(TEST_SGI_ID, 0x1 << 3);
+	tee_time_wait(200);
+	if (memcmp(test_sgi_value, expect_sgi_value, sizeof(test_sgi_value)))
+		return TEE_ERROR_GENERIC;
+
+	expect_sgi_value[0]++;
+	expect_sgi_value[1]++;
+	expect_sgi_value[2]++;
+	expect_sgi_value[3]++;
+	itr_raise_sgi(TEST_SGI_ID, 0xff);
+	tee_time_wait(200);
+	if (memcmp(test_sgi_value, expect_sgi_value, sizeof(test_sgi_value)))
+		return TEE_ERROR_GENERIC;
+
+	expect_sgi_value[2]++;
+	expect_sgi_value[3]++;
+	itr_raise_sgi(TEST_SGI_ID, 0x1 << 2);
+	itr_raise_sgi(TEST_SGI_ID, 0x1 << 3);
+	tee_time_wait(200);
+	if (memcmp(test_sgi_value, expect_sgi_value, sizeof(test_sgi_value)))
+		return TEE_ERROR_GENERIC;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result test_spi(void)
+{
+	itr_add(&spi_handler);
+	itr_enable(TEST_SPI_ID);
+
+	expect_spi_value[3]++;
+	itr_set_affinity(TEST_SPI_ID, 0x1 << 3);
+	itr_raise_pi(TEST_SPI_ID);
+	tee_time_wait(200);
+	if (memcmp(test_spi_value, expect_spi_value, sizeof(test_spi_value)))
+		return TEE_ERROR_GENERIC;
+
+	expect_spi_value[2]++;
+	itr_set_affinity(TEST_SPI_ID, 0x1 << 2);
+	itr_raise_pi(TEST_SPI_ID);
+	tee_time_wait(200);
+	if (memcmp(test_spi_value, expect_spi_value, sizeof(test_spi_value)))
+		return TEE_ERROR_GENERIC;
+
+	expect_spi_value[1]++;
+	itr_set_affinity(TEST_SPI_ID, 0x1 << 1);
+	itr_raise_pi(TEST_SPI_ID);
+	tee_time_wait(200);
+	if (memcmp(test_spi_value, expect_spi_value, sizeof(test_spi_value)))
+		return TEE_ERROR_GENERIC;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result test_ppi(void)
+{
+	itr_add(&ppi_handler);
+	itr_enable(TEST_PPI_ID);
+
```

Comment:
IRQs shall be mask during this test so that its execution does not switch between cpu cores.